File: Sources/Max/Core/Channels/DiscordChannel.swift
import Foundation

/// Discord bot via the Gateway (WebSocket). Connect → HELLO (heartbeat
/// interval) → send heartbeats + IDENTIFY → receive MESSAGE_CREATE dispatches →
/// reply over the REST API. Requires a bot token with the MESSAGE CONTENT
/// privileged intent enabled in the Discord developer portal.
@MainActor
final class DiscordChannel {
    static let shared = DiscordChannel()

    private var task: URLSessionWebSocketTask?
    private var heartbeatTimer: Timer?
    private var seq: Int?
    private var token = ""
    private var running = false

    private init() {}

    func startIfEnabled() {
        let config = MaxConfig.load()
        guard config.discordEnabled,
              let t = SecretStore.secret(forKey: "discord-bot"), !t.isEmpty else {
            stop(); return
        }
        guard !running else { return }
        token = t
        running = true
        connect()
    }

    func stop() {
        running = false
        heartbeatTimer?.invalidate(); heartbeatTimer = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    private func connect() {
        guard running else { return }
        seq = nil
        let url = URL(string: "wss://gateway.discord.gg/?v=10&encoding=json")!
        let t = URLSession.shared.webSocketTask(with: url)
        task = t
        t.resume()
        receive()
    }

    private func reconnect() {
        guard running else { return }
        heartbeatTimer?.invalidate(); heartbeatTimer = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self] in
            guard let self, self.running else { return }
            self.connect()
        }
    }

    private func receive() {
        task?.receive { [weak self] result in
            Task { @MainActor in
                guard let self, self.running else { return }
                switch result {
                case .failure:
                    self.reconnect()
                case .success(let message):
                    switch message {
                    case .string(let s): self.handleFrame(s)
                    case .data(let d): if let s = String(data: d, encoding: .utf8) { self.handleFrame(s) }
                    @unknown default: break
                    }
                    self.receive()
                }
            }
        }
    }

    private func handleFrame(_ raw: String) {
        guard let obj = JSON.decode(raw) else { return }
        if let s = obj["s"] as? Int { seq = s }
        switch obj["op"] as? Int {
        case 10: // HELLO
            let d = obj["d"] as? [String: Any]
            let interval = (d?["heartbeat_interval"] as? Double ?? 41250) / 1000.0
            startHeartbeat(interval)
            identify()
        case 0: // dispatch
            if obj["t"] as? String == "MESSAGE_CREATE", let d = obj["d"] as? [String: Any] {
                onMessage(d)
            }
        case 7, 9: // reconnect / invalid session
            reconnect()
        default:
            break
        }
    }

    private func startHeartbeat(_ interval: TimeInterval) {
        heartbeatTimer?.invalidate()
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self else { return }
                self.send(["op": 1, "d": self.seq as Any])
            }
        }
    }

    private func identify() {
        // GUILD_MESSAGES (1<<9) | DIRECT_MESSAGES (1<<12) | MESSAGE_CONTENT (1<<15)
        let intents = (1 << 9) | (1 << 12) | (1 << 15)
        send(["op": 2, "d": [
            "token": token,
            "intents": intents,
            "properties": ["os": "macOS", "browser": "max", "device": "max"],
        ]])
    }

    private func onMessage(_ d: [String: Any]) {
        let author = d["author"] as? [String: Any] ?? [:]
        if author["bot"] as? Bool == true { return } // ignore bots, including self
        let authorId = author["id"] as? String ?? ""
        let authorName = author["username"] as? String ?? "?"
        let content = (d["content"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let channelId = d["channel_id"] as? String ?? ""
        guard !content.isEmpty, !channelId.isEmpty, !authorId.isEmpty else { return }

        let allow = MaxConfig.load().discordAllowlist
        guard allow.contains(authorId) else {
            // Log the sender so the user can find their ID and allowlist it.
            ChannelLog.write("discord: ignored message from \(authorName) (id \(authorId)) — not in allowlist")
            return
        }

        let token = self.token
        ChannelRouter.shared.handle(key: "discord-\(channelId)", text: content) { reply in
            Self.sendReply(channelId: channelId, text: reply, token: token)
        }
    }

    private func send(_ obj: [String: Any]) {
        task?.send(.string(JSON.encodeString(obj))) { _ in }
    }

    nonisolated static func sendReply(channelId: String, text: String, token: String) {
        var request = URLRequest(url: URL(string: "https://discord.com/api/v10/channels/\(channelId)/messages")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bot \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = JSON.encode(["content": String(text.prefix(1900))])
        URLSession.shared.dataTask(with: request).resume()
    }
}

/// Shared lightweight log for channel diagnostics (sender IDs, errors).
enum ChannelLog {
    static func write(_ line: String) {
        let url = MaxPaths.root.appendingPathComponent("channels.log")
        let stamp = ISO8601DateFormatter().string(from: Date())
        let text = "[\(stamp)] \(line)\n"
        if let h = try? FileHandle(forWritingTo: url) {
            h.seekToEndOfFile(); h.write(Data(text.utf8)); try? h.close()
        } else {
            try? text.write(to: url, atomically: true, encoding: .utf8)
        }
    }
}

File: windows/Max.Core/Channels/DiscordChannel.cs
using System.Net.Http.Headers;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json.Nodes;

namespace Max.Core;

/// <summary>
/// Discord bot channel over the Gateway WebSocket (HELLO → IDENTIFY → heartbeat →
/// MESSAGE_CREATE), replying via the REST API. Needs the MESSAGE CONTENT privileged
/// intent enabled in the Developer Portal. Allowlist = your numeric Discord user ID.
/// </summary>
public sealed class DiscordChannel
{
    private const int Intents = (1 << 9) | (1 << 12) | (1 << 15); // GUILD_MESSAGES | DIRECT_MESSAGES | MESSAGE_CONTENT
    private static readonly HttpClient Http = new();
    private readonly ISecretStore _secrets;
    private readonly ChannelRouter _router;
    private CancellationTokenSource? _cts;

    public DiscordChannel(ISecretStore secrets, ChannelRouter router) { _secrets = secrets; _router = router; }

    public void Start()
    {
        var config = MaxConfig.Load();
        var token = _secrets.Get("discord-bot");
        if (!config.DiscordEnabled || string.IsNullOrEmpty(token)) return;
        if (_cts != null) return;
        _cts = new CancellationTokenSource();
        _ = RunAsync(token!, _cts.Token);
    }

    public void Stop() { _cts?.Cancel(); _cts = null; }

    private async Task RunAsync(string token, CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try { await ConnectOnceAsync(token, ct); }
            catch (OperationCanceledException) { return; }
            catch (Exception ex) { ChannelLog.Write($"discord: {ex.Message}"); }
            try { await Task.Delay(5000, ct); } catch { return; } // reconnect backoff
        }
    }

    private async Task ConnectOnceAsync(string token, CancellationToken ct)
    {
        using var ws = new ClientWebSocket();
        await ws.ConnectAsync(new Uri("wss://gateway.discord.gg/?v=10&encoding=json"), ct);
        var sendLock = new SemaphoreSlim(1, 1);
        int? lastSeq = null;
        var heartbeatCts = CancellationTokenSource.CreateLinkedTokenSource(ct);

        async Task Send(JsonObject payload)
        {
            await sendLock.WaitAsync(ct);
            try
            {
                var bytes = Encoding.UTF8.GetBytes(payload.ToJsonString());
                await ws.SendAsync(bytes, WebSocketMessageType.Text, true, ct);
            }
            finally { sendLock.Release(); }
        }

        ChannelLog.Write("discord: connected, awaiting hello");

        while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
        {
            var raw = await ReceiveAsync(ws, ct);
            if (raw is null) break;
            JsonObject msg;
            try { msg = JsonNode.Parse(raw)!.AsObject(); } catch { continue; }

            var op = (int?)msg["op"] ?? -1;
            if (msg["s"] is JsonNode s && s.GetValueKind() == System.Text.Json.JsonValueKind.Number) lastSeq = (int)s;

            switch (op)
            {
                case 10: // HELLO
                    var interval = (int)msg["d"]!["heartbeat_interval"]!;
                    _ = HeartbeatLoop(Send, () => lastSeq, interval, heartbeatCts.Token);
                    await Send(new JsonObject
                    {
                        ["op"] = 2,
                        ["d"] = new JsonObject
                        {
                            ["token"] = token,
                            ["intents"] = Intents,
                            ["properties"] = new JsonObject { ["os"] = "windows", ["browser"] = "max", ["device"] = "max" },
                        },
                    });
                    break;

                case 0: // dispatch
                    if ((string?)msg["t"] == "MESSAGE_CREATE")
                        await HandleMessageAsync(msg["d"]!.AsObject(), token, ct);
                    break;

                case 7:  // reconnect requested
                case 9:  // invalid session
                    heartbeatCts.Cancel();
                    return;
            }
        }
        heartbeatCts.Cancel();
    }

    private static async Task HeartbeatLoop(Func<JsonObject, Task> send, Func<int?> seq, int intervalMs, CancellationToken ct)
    {
        try
        {
            while (!ct.IsCancellationRequested)
            {
                await Task.Delay(intervalMs, ct);
                await send(new JsonObject { ["op"] = 1, ["d"] = seq() is int n ? n : (JsonNode?)null });
            }
        }
        catch { }
    }

    private async Task HandleMessageAsync(JsonObject d, string token, CancellationToken ct)
    {
        var author = d["author"]?.AsObject();
        var content = (string?)d["content"];
        var channelId = (string?)d["channel_id"];
        if (author is null || string.IsNullOrWhiteSpace(content) || channelId is null) return;
        if (author["bot"]?.GetValue<bool>() == true) return;

        var userId = (string?)author["id"] ?? "";
        var allow = MaxConfig.Load().DiscordAllowlist;
        if (!allow.Contains(userId))
        {
            ChannelLog.Write($"discord: ignored message from {(string?)author["username"] ?? "?"} (id {userId}) — not in allowlist");
            return;
        }

        ChannelLog.Write($"discord: handling message from {userId}");
        await _router.HandleAsync($"discord-{channelId}", content!.Trim(),
            reply => SendMessageAsync(channelId, reply, token), ct);
    }

    private static async Task SendMessageAsync(string channelId, string text, string token)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, $"https://discord.com/api/v10/channels/{channelId}/messages");
            req.Headers.Authorization = new AuthenticationHeaderValue("Bot", token);
            var body = new JsonObject { ["content"] = text.Length > 1900 ? text[..1900] : text };
            req.Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
            await Http.SendAsync(req);
        }
        catch (Exception ex) { ChannelLog.Write($"discord: send error {ex.Message}"); }
    }

    private static async Task<string?> ReceiveAsync(ClientWebSocket ws, CancellationToken ct)
    {
        var buffer = new byte[16 * 1024];
        using var ms = new MemoryStream();
        WebSocketReceiveResult result;
        do
        {
            try { result = await ws.ReceiveAsync(buffer, ct); }
            catch { return null; }
            if (result.MessageType == WebSocketMessageType.Close) return null;
            ms.Write(buffer, 0, result.Count);
        } while (!result.EndOfMessage);
        return Encoding.UTF8.GetString(ms.ToArray());
    }
}
